py/surveyqa/summary.py

```diff
@@ -158,29 +157,26 @@ def get_summarytable(exposures):
 
     Returns a bokeh DataTable object.
     '''
-    night = exposures['NIGHT']
-    sorted_total = OrderedDict(Counter(night))
-    night_exps_total = np.array(sorted_total.values())
-
-    #list of counts of each program for each night
-    dct = []
-    for n in list(sorted_total.keys()):
-        keep = exposures['NIGHT'] == n
-        nights_selected = exposures[keep]
-        program_freq = Counter(nights_selected['PROGRAM'])
-        dct.append(program_freq)
-
-    #get the values out of the dictionaries above (in a specified order so we can splice later)
-    counts = []
-    for d in dct:
-        for key in ['BRIGHT', 'GRAY', 'DARK', 'CALIB']:
-            counts.append(d[key])
-
-    #lists of counts of each program for each night
-    brights = np.array([counts[i] for i in np.arange(0,len(counts), 4)])
-    grays = np.array([counts[i] for i in np.arange(1,len(counts), 4)])
-    darks = np.array([counts[i] for i in np.arange(2,len(counts), 4)])
-    calibs = np.array([counts[i] for i in np.arange(3,len(counts), 4)])
+    nights = np.unique(exposures['NIGHT'])
+
+    isbright = (exposures['PROGRAM'] == 'BRIGHT')
+    isgray = (exposures['PROGRAM'] == 'GRAY')
+    isdark = (exposures['PROGRAM'] == 'DARK')
+    iscalib = (exposures['PROGRAM'] == 'CALIB')
+
+    num_nights = len(nights)
+    brights = list()
+    grays = list()
+    darks = list()
+    calibs = list()
+    totals = list()
+    for n in nights:
+        thisnight = exposures['NIGHT'] == n
+        totals.append(np.count_nonzero(thisnight))
+        brights.append(np.count_nonzero(thisnight & isbright))
+        grays.append(np.count_nonzero(thisnight & isgray))
+        darks.append(np.count_nonzero(thisnight & isdark))
+        calibs.append(np.count_nonzero(thisnight & iscalib))
```

Comment:
Refactored to have a single loop over nights, filter on that night, and then combine with the previously pre-calculated isdark/isgray/isbright/iscalib arrays of booleans, avoiding multiple calls to Counter.  Improvement here was by experimentation and verification, but followed the basic optimization principle of favoring numpy vectorized operations over explicit loops.